File: bound/kt/src/main/kotlin/web5/sdk/crypto/keys/InMemoryKeyManager.kt
package web5.sdk.crypto.keys

import web5.sdk.crypto.signers.Signer
import web5.sdk.crypto.signers.ToOuterSigner
import tbdex.sdk.rust.InMemoryKeyManager as RustCoreInMemoryKeyManager
import tbdex.sdk.rust.KeyManager as RustCoreKeyManager

internal class InMemoryKeyManager : KeyManager {
    private val rustCoreInMemoryKeyManager = RustCoreInMemoryKeyManager()

    constructor(privateJwks: List<Jwk>) {
        privateJwks.forEach {
            this.rustCoreInMemoryKeyManager.importPrivateJwk(it)
        }
    }

    override fun getSigner(publicJwk: Jwk): Signer {
        val innerSigner = this.rustCoreInMemoryKeyManager.getSigner(publicJwk)
        return ToOuterSigner(innerSigner)
    }

    override fun getRustCoreKeyManager(): RustCoreKeyManager {
        return this.rustCoreInMemoryKeyManager.getAsKeyManager()
    }

    fun importPrivateJwk(privateJwk: Jwk): Jwk {
        return this.rustCoreInMemoryKeyManager.importPrivateJwk(privateJwk)
    }
}

File: bound/typescript/src/crypto/in-memory-key-manager.ts
import wasm from "../wasm";
import { Jwk } from "./jwk";
import { KeyManager } from "./key-manager";
import { Signer } from "./signer";

export class InMemoryKeyManager implements KeyManager {
  private readonly wasmKeyManager: wasm.WasmKeyManager;

  constructor() {
    this.wasmKeyManager = wasm.new_in_memory_key_manager();
  }

  importPrivateJwk(privateJwk: Jwk): Jwk {
    const privateJwkJson = JSON.stringify(privateJwk);
    const publicJwkJson =
      this.wasmKeyManager.import_private_jwk(privateJwkJson);
    const publicJwk: Jwk = JSON.parse(publicJwkJson);
    return publicJwk;
  }

  getSigner(publicJwk: Jwk): Signer {
    const publicJwkJson = JSON.stringify(publicJwk);
    const wasmSigner = this.wasmKeyManager.get_signer(publicJwkJson);
    return Signer.fromWASM(wasmSigner);
  }
}
